search/interpolation_search.cpp

```diff
@@ -1,61 +1,123 @@
+namespace interpolation_search {
 
-    std::cout << "Enter the size of array(less than 100) : ";
-    std::cin >> n;
+/******************************************************************************
+ * @brief The main function which implements interpolation search
+ * @param arr vector to be searched in
+ * @param number value to be searched
+ * @returns integer index of `number` in vector `arr`
+ *******************************************************************************/
+uint64_t interpolationSearch(const std::vector<uint64_t> &arr,
+                             uint64_t number) {
+    uint64_t size = arr.size();
+    uint64_t low = 0, high = (size - 1);
 
-    int *array = new int[n];
+    // Since vector is sorted, an element present in array must be in range
+    // defined by corner
+    while (low <= high && number >= arr[low] && number <= arr[high]) {
+        if (low == high) {
+            if (arr[low] == number) {
+                return low;
+            }
+            return -1;
+        }
+        // Probing the position with keeping uniform distribution in mind.
+        uint64_t pos =
+            low +
+            ((static_cast<uint64_t>(high - low) / (arr[high] - arr[low])) *
+             (number - arr[low]));
 
-    std::cout << "array in ascending (increasing) order : " << std::endl;
+        if (arr[pos] == number) {
+            return pos;  // Condition of target found
+        }
 
-    for (int i = 0; i < n; i++) std::cin >> array[i];
+        if (arr[pos] < number) {
+            low = pos + 1;  // If x is larger, x is in upper part
+        }
 
-    std::cout << "Enter the value you want to search : ";
-    std::cin >> value;
+        else {
+            high = pos - 1;  // If x is smaller, x is in the lower part
+        }
+    }
+    return -1;
+}
 
-    re = interpolation_search(array, value, n);
+}  // namespace interpolation_search
 
-    if (re == -1)
-        std::cout << "Entered value is not in the array" << std::endl;
-    else
-        std::cout << "The value is at the position " << re << std::endl;
+}  // namespace search
 
-    delete[] array;
+/*******************************************************************************
+ * @brief Self-test implementation
+ * @returns void
+ *******************************************************************************/
+static void tests() {
+    // testcase
+    // array = [10, 12, 13, 16, 18, 19, 20, 21, 1, 2, 3, 4, 22, 23, 24, 33, 35,
+    // 42, 47] , Value = 33 should return 15
+    std::vector<uint64_t> arr = {{10, 12, 13, 16, 18, 19, 20, 21, 1, 2, 3, 4,
+                                  22, 23, 24, 33, 35, 42, 47}};
+    sort(arr.begin(), arr.end());
+    uint64_t number = 33;  // Element to be searched
+    uint64_t expected_answer = 15;
+    uint64_t derived_answer =
+        search::interpolation_search::interpolationSearch(arr, number);
+    std::cout << "Testcase : ";
+    assert(derived_answer == expected_answer);
+    std::cout << "Passed!";
```

Comment:
Let's add a new line here. It'll look better in the terminal. 😃
```suggestion
    std::cout << "Passed!\n";
```